### PR TITLE
Support sparse tensors in THPP

### PR DESCRIPTION
**Pull request by colesbury**
The only op currently supported is dense tensor + sparse tensor addition, which is all we need for autograd.

I've made `tensor_type tensor` public. The alternative is to add a getter, but getters always seemed verbose and lame in C++.

### CODE REVIEW

File: torch/lib/THPP/tensors/THCTensor.hpp [hunk 1]
```diff
@@ -149,7 +150,7 @@ struct THCTensor : public interface_traits<real>::tensor_interface_type {
   THCTensor& resize(const iterator& size_begin, const iterator& size_end,
                    const iterator& stride_begin, const iterator& stride_end);
 
-protected:
+public:
```
Comment (apaszke) [inline torch/lib/THPP/tensors/THCTensor.hpp]: isn't it enough to remove this?

File: torch/lib/THPP/tensors/generic/THSTensor.cpp [hunk 2]
```diff
@@ -0,0 +1,454 @@
...
+template<>
+auto THSTensor<real>::transpose(const Tensor& src, int dimension1,
+                                int dimension2) -> THSTensor& {
+  throw std::runtime_error("THSTensor::&() not supported");}
+
+
+template<>
+auto THSTensor<real>::unfold(const Tensor& src, int dimension,
+                             long size, long step) ->THSTensor& {
+  throw std::runtime_error("THSTensor::)() not supported");}
```
Comment (apaszke) [inline torch/lib/THPP/tensors/generic/THSTensor.cpp]: :)()

**Comment by apaszke** [comment]
I've merged my branch and we have a conflict now 😕

File: torch/lib/THPP/tensors/THCTensor.hpp (hunk 1 above)
Comment (colesbury) [inline torch/lib/THPP/tensors/THCTensor.hpp]: There's a `private:` above for resize and it seems nice to group functions and variables in the struct definition.

File: torch/lib/THPP/tensors/generic/THCTensor.cpp [hunk 3]
```diff
@@ -605,6 +605,11 @@ bool THCTensor<real>::isCuda() const {
 }
 
 template<>
+bool THCTensor<real>::isSparse() const {
+  return true;
```
Comment (fmassa) [inline torch/lib/THPP/tensors/generic/THCTensor.cpp]: Is this supposed to be true?

Comment (colesbury) [inline torch/lib/THPP/tensors/generic/THCTensor.cpp]: oops

Comment (colesbury) [inline torch/lib/THPP/tensors/generic/THCTensor.cpp]: Fixed

File: torch/lib/THPP/tensors/generic/THSTensor.cpp [hunk 4]
```diff
@@ -0,0 +1,454 @@
...
+}
+
+template<>
+std::size_t THSTensor<real>::elementSize() const {
+  return sizeof(real);
+}
+
+template<>
+long long THSTensor<real>::numel() const {
+  throw std::runtime_error("THSTensor::>::() not supported");
```
Comment (fmassa) [inline torch/lib/THPP/tensors/generic/THSTensor.cpp]: nit: There are a couple of bad replacements in the following functions:
* `data` (both of them)
* `cdata` (both)

Comment (colesbury) [inline torch/lib/THPP/tensors/generic/THSTensor.cpp]: My editor magic failed me :(

**Pull request merged**